katdal/applycal.py

```diff
@@ -0,0 +1,188 @@
+                     for (ant_idx, ant) in enumerate(cal_ants)}
+    if not cal_input_map:
+        return
+
+    def calc_correction_per_input(cache, name, inp, product):
+        """Calculate correction sensor for input `inp` from cal solutions."""
+        try:
+            index = cal_input_map[inp]
+        except KeyError:
+            raise KeyError("No calibration solutions available for input "
+                           "'{}' - available ones are {}"
+                           .format(inp, sorted(cal_input_map.keys())))
+        if product.startswith('K'):
+            sensor_data = calc_delay_correction(cache, product, index, freqs)
+        else:
+            raise KeyError("Unknown calibration product '{}'".format(product))
+        cache[name] = sensor_data
+        return sensor_data
+
+    correction_sensor = 'Calibration/{inp}_correction_{product}'
+    cache.virtual[correction_sensor] = calc_correction_per_input
+
+
+def calc_correction_per_corrprod(dump, channels, cache, inputs,
+                                 input1_index, input2_index, cal_products):
+    """Gain correction per channel per correlation product for a given dump.
+
+    This calculates an array of complex gain correction terms of shape
+    (n_chans, n_corrprods) that can be directly applied to visibility data.
+    This incorporates all requested calibration products at the specified
+    dump and channels.
+
+    Parameters
+    ----------
+    dump : int
+        Dump index (applicable to full data set, i.e. absolute)
+    channels : list of int, length n_chans
+        Channel indices (applicable to full data set, i.e. absolute)
+    cache : :class:`katdal.sensordata.SensorCache` object
+        Sensor cache, used to look up individual correction sensors
+    inputs : sequence of string
+        Correlator input labels
+    input1_index, input2_index : list of int, length n_corrprods
+        Indices into `inputs` of first and second items of correlation product
+    cal_products : sequence of string
+        Calibration products that will contribute to corrections
+
+    Returns
+    -------
+    gains : array of complex64, shape (n_chans, n_corrprods)
+        Gain corrections per channel per correlation product
+
+    Raises
+    ------
+    KeyError
+        If input and/or cal product has no associated correction
+    """
+    g_per_input = np.ones((len(inputs), len(channels)), dtype='complex64')
+    for product in cal_products:
+        for n, inp in enumerate(inputs):
+            sensor_name = 'Calibration/{}_correction_{}'.format(inp, product)
+            g_product = cache.get(sensor_name)[dump]
+            if np.shape(g_product) != ():
+                g_product = g_product[channels]
+            g_per_input[n] *= g_product
+    g_per_cp = g_per_input[input1_index] * g_per_input[input2_index].conj()
+    return g_per_cp.T
+
+
+def apply_vis_correction(out, correction):
+    """Apply `correction` in-place to visibility data in `out`."""
```

Comment:
The latter.

This function group encapsulates the differences in applying corrections to vis, weights and flags. Here is a rough sketch:

- **Vis:**
   NaN in correction => 1 (don't destroy)
   out = out * correction

- **Weights:**
   NaN in correction -> 0 (kill the weight)
   out = out / (|correction| ** 2)

- **Flags:**
   NaN in correction -> add cal_failure flag to out

Reply:
In that case it definitely shouldn't be modifying the correction in-place, unless copying is being done at a higher level, since doing it here will destroy the information about where the NaNs are.

The cheapest option might be something like
```python
np.multiply(out, correction, out=out, where=~np.isnan(correction))
```
which avoids creating any temporary arrays of complex numbers. But that's a guess not a benchmark.

Reply:
Copying is being done at a higher level...

I picture the following in `VisibilityDataV4._set_keep`:

```Python
add_applycal_transform(f.vis, ..., apply_vis_correction)
add_applycal_transform(f.weights, ..., apply_weight_correction)
add_applycal_transform(f.flags, ..., apply_flag_correction)
```
This is the appropriate scope to identify each indexer type and apply the relevant correction. I want `calc_correction_per_corrprod` to be shared by all three. Each indexer dask graph therefore gets its own `map_blocks` with a custom `calibrate_chunk` function, and corrections are not shared amongst the three indexers. This seems wasteful but at least takes care of the different chunking schemes in each dask array.